sway/commands/resize.c

```diff
@@ -130,177 +130,108 @@ static bool is_horizontal(uint32_t axis) {
-	int major_weight = 0;
-	while (parent) {
-		list_t *siblings = container_get_siblings(parent);
-		if (container_parent_layout(parent) == parallel_layout) {
-			for (int i = 0; i < siblings->length; i++) {
-				struct sway_container *sibling = siblings->items[i];
-
-				int sibling_pos = parallel_coord(sibling, axis);
-				int focused_pos = parallel_coord(focused, axis);
-				int parent_pos = parallel_coord(parent, axis);
-
-				if (sibling_pos != focused_pos) {
-					if (sibling_pos < parent_pos) {
-						minor_weight++;
-					} else if (sibling_pos > parent_pos) {
-						major_weight++;
-					}
-				}
-			}
-			if (major_weight || minor_weight) {
-				break;
-			}
-		}
-		parent = parent->parent;
-	}
-	if (!parent) {
+	con = container_find_resize_parent(con, axis);
+	if (!con) {
 		// Can't resize in this direction
 		return;
 	}
 
-	// Implement up/down/left/right direction by zeroing one of the weights
-	if (axis == WLR_EDGE_TOP || axis == WLR_EDGE_LEFT) {
-		major_weight = 0;
-	} else if (axis == WLR_EDGE_RIGHT || axis == WLR_EDGE_BOTTOM) {
-		minor_weight = 0;
-	}
-
-	bool horizontal = is_horizontal(axis);
-	int min_sane = horizontal ? MIN_SANE_W : MIN_SANE_H;
-
-	//TODO: Ensure rounding is done in such a way that there are NO pixel leaks
-	// ^ ?????
-	list_t *siblings = container_get_siblings(parent);
-
-	for (int i = 0; i < siblings->length; i++) {
-		struct sway_container *sibling = siblings->items[i];
-
-		int sibling_pos = parallel_coord(sibling, axis);
-		int focused_pos = parallel_coord(focused, axis);
-		int parent_pos = parallel_coord(parent, axis);
-
-		int sibling_size = parallel_size(sibling, axis);
-		int parent_size = parallel_size(parent, axis);
-
-		if (sibling_pos != focused_pos) {
-			if (sibling_pos < parent_pos && minor_weight) {
-				double pixels = -amount / minor_weight;
-				if (major_weight && (sibling_size + pixels / 2) < min_sane) {
-					return; // Too small
-				} else if (!major_weight && sibling_size + pixels < min_sane) {
-					return; // Too small
-				}
-			} else if (sibling_pos > parent_pos && major_weight) {
-				double pixels = -amount / major_weight;
-				if (minor_weight && (sibling_size + pixels / 2) < min_sane) {
-					return; // Too small
-				} else if (!minor_weight && sibling_size + pixels < min_sane) {
-					return; // Too small
-				}
-			}
-		} else {
-			double pixels = amount;
-			if (parent_size + pixels < min_sane) {
-				return; // Too small
-			}
+	// For HORIZONTAL or VERTICAL, we are growing in two directions so select
+	// both adjacent siblings. For RIGHT or DOWN, just select the next sibling.
+	// For LEFT or UP, convert it to a RIGHT or DOWN resize and reassign con to
+	// the previous sibling.
+	struct sway_container *prev = NULL;
+	struct sway_container *next = NULL;
+	list_t *siblings = container_get_siblings(con);
+	int index = container_sibling_index(con);
+
+	if (axis == AXIS_HORIZONTAL || axis == AXIS_VERTICAL) {
+		prev = siblings->items[index - 1];
```

Comment:
When running `resize grow width 10px` or `resize grow height 10px` on the first or last child, sway segfaults since `prev` and `next` get assigned invalid values here that are later dereferenced.

Reply:
Fixed.